ftw/solr/tests/test_collective_indexing_integration.py

```diff
@@ -141,6 +141,73 @@ def test_reindex_object_security_is_recursive(self):
         # XXX: Why isn't call order stable here?
         self.connection.add.assert_has_calls(expected_calls, any_order=True)
 
+    def test_system_roles_dont_mistakenly_terminate_recursion(self):
+        """This test makes sure that the special shortcut treatment of the
+        'Authenticated' and 'Anonymous' roles in the allowedRolesAndUsers
+        indexer doesn't mistakenly stop down propagation of security
+        reindexing in our reindexObjectSecurity patch.
+        """
+        # Prepare a situation as follows:
+        #
+        # - folder2 has the View permission assigned to Authenticated and
+        #   Reader roles
+        # - Its child, subfolder2_without_aq, only has View assigned to Reader,
+        #   and it doesn't acquire the View permission
+        # - Give the user the local role Reader on the container folder2
+        # - Because of local role inheritance, the user also has Reader on
+        #   subfolder2_without_aq. This is the *only* setting by which he
+        #   gets the 'View' permission on subfolder2_without_aq
+        self.folder2.manage_permission(
+            'View', roles=['Authenticated', 'Reader'], acquire=False)
+        self.subfolder2_without_aq.manage_permission(
+            'View', roles=['Reader'], acquire=False)
+
+        self.folder2.manage_setLocalRoles(TEST_USER_ID, ['Reader'])
+
+        # Make sure the user doesn't have any local roles on
+        # subfolder2_without_aq, just to be sure
+        self.subfolder2_without_aq.manage_delLocalRoles(TEST_USER_ID)
```

Comment:
as an alternative you could document this pre-condition with an assertion 🤔

Reply:
👍 updated.